src/puppetlabs/puppetdb/cli/export.clj

```diff
@@ -46,213 +67,103 @@
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;; General functions
 
-(defn parse-response
-  "The parsed JSON response body"
-  [{:keys [status body]}]
-  (when (= status 200)
-    (seq (json/parse-string body true))))
-
-(pls/defn-validated retrieve-related-data
-  "Given a base-url and a specific URL retrieve the related data"
-  [base-url :- utils/base-url-schema
-   url :- s/Str]
-  (parse-response
-   (client/get
-    (str (utils/base-url->str-no-path base-url) url))))
-
-(def expanded-schema
-  "Expanded schema"
-  {:data s/Any
-   :href s/Str})
-
-(def unexpanded-schema
-  "Unexpanded schema"
-  {(s/optional-key :data) s/Any
-   :href s/Str})
-
-(pls/defn-validated complete-unexpanded :- expanded-schema
+(pls/defn-validated complete-unexpanded-fields
   "Complete the unexpanded data, by retrieving data from the href if data is
   missing"
-  [{:keys [data href] :as value :- unexpanded-schema}
-   base-url :- utils/base-url-schema]
-  (if data
-    value
-    (assoc value :data (retrieve-related-data base-url href))))
+  [query-fn
+   fields :- [s/Keyword]
+   values :- [{s/Any s/Any}]]
+  (->> values
+       (map (partial
+             kitchensink/mapvals
+             #(utils/assoc-when % :data (query-child-href-internally query-fn (:href %)))
+             fields))))
```

Comment:
Is there a way to write the above more simply? Maybe as 2 fns or something, or by naming some of the intermediate values. I'm having a hard time working out what it does.